### PR TITLE
Split librustc::{traits,infer} to their respective crates

### PR DESCRIPTION
**Pull request by cjgillot**
Followup on #67953.

I tried to follow the existing module structures.

cc @eddyb  
r? @Zoxc

### CODE REVIEW

**Comment by Zoxc** [comment]
Would `rustc_trait_selection` be a better crate name for `rustc_infer::traits`?

cc @rust-lang/compiler

**Comment by Zoxc** [comment]
Could you split `Querify object_safety_violations.` into a separate PR so we can run perf on it?

**Comment by Zoxc** [comment]
This seems to be a bit too heavy on trait stuff for me to review.

r? @matthewjasper

File: src/librustc_infer/lib.rs [hunk 1]
```diff
@@ -17,12 +16,10 @@
 #![feature(bool_to_option)]
 #![feature(box_patterns)]
 #![feature(box_syntax)]
-#![feature(drain_filter)]
 #![feature(never_type)]
 #![feature(range_is_empty)]
 #![feature(in_band_lifetimes)]
 #![feature(crate_visibility_modifier)]
-#![recursion_limit = "512"]
```
Comment (matthewjasper) [inline src/librustc_infer/lib.rs]: This was only needed for the parallel compiler. Can you try building it before removing this?

Comment (cjgillot) [reply src/librustc_infer/lib.rs]: I managed to compile as is.

**Comment by petrochenkov** [comment]
>ping in https://github.com/rust-lang/rust/pull/69076#issuecomment-586948423

[triagebot] Not my area of expertise.

**Comment by RalfJung** [comment]
> +5,583 −4,693

Curious that what should be mostly a move adds almost 1000 LOC.

**Comment by cjgillot** [comment]
Rebased.

Moving from inherent methods to extension traits generates a lot of boilerplate.

**Comment by RalfJung** [comment]
Ah I see, makes sense.
At some point I plan to play with https://crates.io/crates/extension-trait, Miri has a similar issue.

File: src/librustc_infer/lib.rs (hunk 1 above)
Comment (Zoxc) [reply src/librustc_infer/lib.rs]: Looks like this is needed anyway for `rustdoc`.

**Comment by cjgillot** [comment]
Rebased.

**Comment by pietroalbini** [comment]
@bors treeclosed=1000

Our CI capacity was reduced, we contacted Microsoft to ask for a fix.

**Comment by pietroalbini** [comment]
@bors retry

Killed the wrong build :sweat_smile:

**Comment by cjgillot** [comment]
Rebased.

**Comment by nnethercote** [comment]
This change [regressed](https://perf.rust-lang.org/compare.html?start=0f1e814c119afe08abff199895f3279ee0376d70&end=131772c5e0ba40cd656dedb5e1990d36e3ea31cf&stat=instructions:u) compile time of the `keccak` benchmark significantly.

@cjgillot: any idea why, and how to fix it?

**Comment by Zoxc** [comment]
@nnethercote It's probably due to things not being inlined anymore. [This PR](https://github.com/rust-lang/rust/pull/70550) seems to [restore performance](https://perf.rust-lang.org/compare.html?start=4911572b2da65eeff0b017831614ee32b546b9db&end=72176503fc0d08a32adfdda3aea038c0f6561a0e) for `keccack`.

**Pull request merged**